### DISCUSSION TITLE
Rootless podman with own network namespace or networking

### DISCUSSION THREAD

**Discussion by electrofloat** [containers/podman#17235] [Q&A]
Hi!

I'm trying to find a way where I can start a rootless podman container and attach it to an already existing netns and/or interface.
I can create the required interfaces and set them up the way I want as root, then I'd like to use those from rootless podman. (maybe I configure 1 bridge and some containers would be part of it so that they can communicate with each-other, maybe I want a different interface for every container with different ip addresses, route them/firewall them the way I want)

The approach written here is almost ok: https://lists.podman.io/archives/list/podman@lists.podman.io/thread/W6MCYO6RY5YFRTSUDAOEZA7SC2EFXRZE/

but it depends on starting the container first (and also depending on it not stopping immediately because network is not available at that point), and also needs to be done every time the container stopped/started. Nor really convenient.

Also here: https://github.com/containers/podman/issues/13706#issuecomment-1083131272 - rhatdan says: "Allowing users to setup their own network namespace and allow that to joined to a Podman container is the best we can do."

Can this already be achieved with current config options for rootless podman, or is this something that needs development?

**Comment by Luap99** [comment]
If you have an netns you can use `podman run --network ns:/path/to/netns ...`. However in order for a rootless user to have permissions to use it the netns must be created by the user within the podman user namespace, then you can use root to add devices into this netns, see https://github.com/containers/podman/issues/10384

**Reply by electrofloat** [reply]
I read #10384 but I don't exactly understand how I would automate this. What I need is: after a system reboot, some script sets up the interfaces / firewall rules etc. as root, then rootless podman containers are started as a user. (and of course the containers would use the already set up networking so no need for slirp4netns, etc.)

From your answer it seems that before anything can happen a podman unshare unshare -n command needs to be executed, but as soon as I exit from that the netns also destroyed. (since that only was associated with that process)

**Reply by Luap99** [reply]
Just run sleep in it and start the process in the background, i.e
`podman unshare unshare -n sh -c "echo $$ > /tmp/netnspid && sleep inf" &` then you can access the pid as root and use nsenter on it to set it up.

Completely untested btw. but I think that should do it.

**Reply by electrofloat** [reply]
That could theoretically work, I haven't tested it yet either, but still feels hacky.
Shouldn't there be some way built into podman to make this more convenient? This is how I imagine it:
1. Run `podman gimme_new_user_ns --user username --name whatever`  as root- This would create the new usernetworknamespace, persist it, write it to a file in a location where the user itself can work with it (like for example where container.conf or network.conf exists for the user with rootless podman), also there should be other commands to list/rm/inspect/rename/whatnot this network namespace. Maybe make it a subcommand/parameter of an already existing podman network
2. Setup the network / firewall the way I want. For example `... ip link set veth300p netns ``podman gimme_the_user_ns --username username --name whatever`` ` still as root
3. Simply start your containers with `podman run --network whatever` - whatever is the name of the usernetworkns

**Reply by Luap99** [reply]
This is not something podman should ever support IMO. Switching users and namespaces would require a lot of fork/exec anyway so there is little to no benefit of having this in podman vs a shell script.
Also consider that this has likely security implications. 
Is there any specific reason why you do not want to use slirp4netns or with v4.4 [pasta](https://passt.top)?

And why not just run podman as root with `--userns=auto` if you need rootful networking anyway?

**Reply by electrofloat** [reply]
I need the network inside a container work like it would on host, but the containers being on a different subnet, with separate ip addresses and no performance impact. As I understand slirp4netns makes it so that the container does not see ip addresses as it would if it had proper network stack.

If I run the container as a user then I don't care if a container is really badly written and runs everything inside it as root, since even if they manage to escape it they are still just the user I run them.

The above also should have the benefit of them being able to bind to ports less than 1024 without any magic.

I also don't yet know pasta, so I have to read on it what it is and why there are two options for rootless networking instead of just one doing everything it should do.

What security implications do you mean? A usernamespace having rootns network? Is that a security issue?

`--userns=auto` still runs the container as root isn't it?

Having the above in podman itself is just for convenience, but maybe I just don't yet understand something about it running container in rootless (for security) and having proper networking (for proper networking).

**Reply by Luap99** [reply]
> What security implications do you mean? A usernamespace having rootns network? Is that a security issue?

I am not saying this is a problem but it could be. It certainly increases the attack surface, all depends on what security issues are in the kernel and how they can be abused.

> --userns=auto still runs the container as root isn't it?

No it creates a user namespace where uid 0 is not mapped, see https://www.redhat.com/sysadmin/rootless-podman-user-namespace-modes

**Reply by electrofloat** [reply]
I see, but that would mean containers creating files on the mounted volume with whatever uid:gid was choosen and that probably could only be accessed by root, which is not good for me. Although now thinking about it maybe I could just use --uidmap to map it to a user I would run it in rootless mode?

So are you telling me that if I run podman as root and with --uidmap, is it exactly the same as running it in rootless mode just with rootns networking?

Also I need podman to **not** touch in any way/shape or form iptables/network interfaces. I do those myself, podmans job is just to run the container with the above restrictions. (I still had to create a "fake" package which provides a totally empty iptables package because on ubuntu 22.04 installing podman automatically installs iptables which is the worst when one is only using nftables, and even then I don't want podman or anything else on the system to touch my firewall config)

**Reply by Luap99** [reply]
> Although now thinking about it maybe I could just use --uidmap to map it to a user I would run it in rootless mode?

Yes this will work.

> So are you telling me that if I run podman as root and with --uidmap, is it exactly the same as running it in rootless mode just with rootns networking?

No of course not, in one case podman runs with user privileges and in the other with root. However in both cases the container will run in a usernamespace where the real root is not mapped.


> Also I need podman to not touch in any way/shape or form iptables/network interfaces.

If you run all containers with `--network ns:...` podman will never touch anything network related. Only when you use named networks it will start configure things accordingly.

Podman works fine with the iptables-nft backend so there is no need for the legacy iptables to be installed.

**Reply by electrofloat** [reply]
The thing is that I don't want iptables binary (let it be legacy or -nft) be on my system at all. nftables exists for about 10 years now, podman should already support it. Not to mention that using rootless podman only, does not need iptables package at all.

Anyways.. I think I'll wait for the 4.4 release (shouldn't it be that far away since 4.4.0-RC3 has just been released the other day, should it?) and also for an official way to install it on ubuntu 22.04, preferably without the words, WIP, CAUTION, UNSTABLE, etc. and try out the new pasta driver.

**Reply by Luap99** [reply]
Well installing dependencies is up to your package manager.

> Not to mention that using rootless podman only, does not need iptables package at all.

Not true, you can use named networks as rootless and this will use CNI or netavark in a extra network namespace and it will call iptables there. It works like that since at least v3.2.

**Reply by mailinglists35** [reply]
is this all (attaching a netns to a rootless) documented in a structured way somewhere, or is everything spread into small parts inside github discussions and issues like here? for example information discussed here and implemented in https://github.com/containers/podman/pull/9423 is not found in https://github.com/containers/podman/blob/main/docs/tutorials/basic_networking.md nor https://www.redhat.com/sysadmin/container-networking-podman

**Reply by daiaji** [reply]
It sounds really complicated and lacks detailed guidelines.
